src/libraries/System.Private.CoreLib/src/System/Reflection/ModifiedType.cs

```diff
@@ -0,0 +1,215 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics.CodeAnalysis;
+using CultureInfo = System.Globalization.CultureInfo;
+
+namespace System.Reflection
+{
+    /// <summary>
+    /// Base class for modified types and standalone modified type.
+    /// Design supports code sharing between different runtimes and lazy loading of custom modifiers.
+    /// </summary>
+    internal partial class ModifiedType : Type
+    {
+        private readonly TypeSignature _typeSignature;
+        private readonly Type _unmodifiedType;
+
+        internal ModifiedType(Type unmodifiedType, TypeSignature typeSignature)
+        {
+            _unmodifiedType = unmodifiedType;
+            _typeSignature = typeSignature;
+        }
+
+        /// <summary>
+        /// Factory to create a node recursively based on the underlying, unmodified type.
+        /// A type tree is formed due to arrays and pointers having an element type, function pointers
+        /// having a return type and parameter types, and generic types having argument types.
+        /// </summary>
+        protected static Type Create(Type unmodifiedType, TypeSignature typeSignature)
+        {
+            Type modifiedType;
+            if (unmodifiedType.IsFunctionPointer)
+            {
+                modifiedType = new ModifiedFunctionPointerType(unmodifiedType, typeSignature);
+            }
+            else if (unmodifiedType.HasElementType)
+            {
+                modifiedType = new ModifiedHasElementType(unmodifiedType, typeSignature);
+            }
+            else if (unmodifiedType.IsGenericType)
+            {
+                modifiedType = new ModifiedGenericType(unmodifiedType, typeSignature);
+            }
+            else
+            {
+                modifiedType = new ModifiedType(unmodifiedType, typeSignature);
+            }
+            return modifiedType;
+        }
+
+        protected Type UnmodifiedType => _unmodifiedType;
+
+        // Below are the multitude of Type overloads. We throw NSE for members that would return an unmodified Type
+        // directly or indirectly. We do this in case we want to support returning modified types instead.
+
+        public override Type[] GetRequiredCustomModifiers()
+        {
+            // No caching is performed; as is the case with FieldInfo.GetCustomModifiers and friends.
+            return GetCustomModifiers(required: true);
+        }
+
+        public override Type[] GetOptionalCustomModifiers()
+        {
+            // No caching is performed; as is the case with FieldInfo.GetCustomModifiers and friends.
+            return GetCustomModifiers(required: false);
+        }
+
+        // Modified types do not support Equals other than when references are equal.
```

Comment:
I am wondering whether it would be better to throw from Equals and GetHashCode instead.

I do not think that this Equals implementation can be ever used for anything useful, and it is not what one would expect. Throwing would make it obvious.

Reply:
I was wondering if there as a precedence for this, and found there is:
https://github.com/dotnet/runtime/blob/main/src/coreclr/nativeaot/System.Private.CoreLib/src/System/TypedReference.cs#L73
https://github.com/dotnet/runtime/blob/main/src/libraries/System.Net.Http/src/System/Net/Http/Headers/HeaderDescriptor.cs#L56
https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/HashCode.cs#L503
https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/ReadOnlySpan.cs#L190
https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Span.cs#L196

So throwing is the safest option at this point; we can of course implement the full Equals if necessary, but I just don't see the value in doing that now.

Reply:
It turns out MetadataLoadContext does use Equals\GetHashCode for caching; determining options.

Reply:
Does it mean that the cache is broken or ineffective in the current version of the PR?

Reply:
Ineffective; changing code to not cache for modified types and ensuring test coverage.